doc/feature/synapses.md

```diff
@@ -0,0 +1,240 @@
+API in Brain to support various common higher-level uses cases, like semantic
+synapse object access or full-circuit synapse density voxelization, to name the
+two extremes.
+
+## Requirements
+
+* Semantic synapse objects that give read-access to all properties
+* Array access which allows vectorization
+* Filtering of synapses for pathway use case (projected synapses of two targets)
+* Lazy-loading to support out-of-core scenarios
+* User-transparent KVS caching
+
+## New dependencies
+
+None
+
+## API
+
+```cpp
+
+namespace brain
+{
+    /*
+     * Synapse GID.
+     *
+     * The GID of a synapse is the a tuple of two numbers:
+     * - The GID of the post-synaptic cell.
+     * - The index of the synapse in the array of afferent contacts
+     *   of the post-synaptic cell before pruning/filtering.
+     * GIDs are invariant regardless of how the structural touches are
+     * converted into functional synapses during circuit building.
+     */
+    typedef std::pair< uint32_t, size > SynapseGID;
+
+    // loading of data during SynapseStream.read(), otherwise load happens on
+    // first touch
+    enum SynapsesPrefetch
+    {
+        SP_NONE = 0,            // only loads pre- and post GIDs
+        SP_ATTRIBUTES = 1 << 0, // topological information (section, segment,
+                                // distance) and electrical attributes
+        SP_POSITIONS = 1 << 1,  // pre/post surface/center positions
+        SP_ALL = SP_ATTRIBUTES | SP_POSITIONS
+    }
+}
+
+class Circuit
+{
+    // max count for read is gids.size()
+    SynapsesStream getAfferentSynapses( const GIDSet& gids,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+
+    // max count for read is gids.size()
+    SynapsesStream getEfferentSynapses( const GIDSet& gids,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+
+    // max count for read is min(preGIDs.size(), postGIDs.size())
+    SynapsesStream getProjectedSynapses( const GIDSet& preGIDs,
+                                         const GIDSet& postGIDs,
+                                     SynapsesPrefetch prefetch = SP_ALL ) const;
+};
+
+// read-only, moveable proxy object of data storage in Synapses container
+class Synapse
+{
+    uint32_t preGID() const;
+    uint32_t preSectionID() const;
+    uint32_t preSegmentID() const;
+    float preDistance() const;
+    Vector3f preSurfacePosition() const;
+    Vector3f preCenterPosition() const;
+
+    uint32_t postGID() const;
+    uint32_t postSectionID() const;
+    uint32_t postSegmentID() const;
+    float postDistance() const;
+    Vector3f postSurfacePosition() const;
+    Vector3f postCenterPosition() const;
+
+    SynapseGID gid() const;
+    float conductance() const;
+    float utilization() const;
+    float depression() const;
+    float facilitation() const;
+    float decay() const;
+    int efficacy() const;
+};
+
+// read-only, moveable, non-copyable, thread-safe
+class Synapses
```

Comment:
Since it's read-only maybe it could be made copyable.
How is the moveable property implemented, C++11?